File: app/src/main/java/com/qimian233/ztool/hook/modules/gametool/SocTemperatureFix.java
package com.qimian233.ztool.hook.modules.gametool;

import com.qimian233.ztool.hook.base.BaseHookModule;
import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XposedHelpers;
import de.robv.android.xposed.callbacks.XC_LoadPackage;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;

/**
 * SOC温度修复Hook模块
 * 功能：拦截游戏服务的温度读取方法，从thermal_zone9文件获取真实温度值
 */
public class SocTemperatureFix extends BaseHookModule {

    private static final String THERMAL_FILE_PATH = "/sys/class/thermal/thermal_zone9/temp";

    @Override
    public String getModuleName() {
        return "Fix_SocTemp";
    }

    @Override
    public String[] getTargetPackages() {
        return new String[]{
                "com.zui.game.service",
                "com.lenovo.gamingservice",
                "com.android.gaming"
        };
    }

    @Override
    public void handleLoadPackage(XC_LoadPackage.LoadPackageParam lpparam) throws Throwable {
        String packageName = lpparam.packageName;

        if (DEBUG) log("SocTemperatureFix: 开始处理包 " + packageName);

        if ("com.zui.game.service".equals(packageName)) {
            hookZuiGameService(lpparam);
        } else if ("com.lenovo.gamingservice".equals(packageName)) {
            hookLenovoGamingService(lpparam);
        } else {
            // 通用游戏服务的Hook
            hookGenericGamingService(lpparam);
        }
    }

    private void hookZuiGameService(XC_LoadPackage.LoadPackageParam lpparam) {
        try {
            Class<?> hwDataInterfaceClass = XposedHelpers.findClass(
                    "com.zui.game.service.util.HWDataInterface",
                    lpparam.classLoader
            );

            // Hook getTemp 方法
            XposedHelpers.findAndHookMethod(
                    hwDataInterfaceClass,
                    "getTemp",
                    new XC_MethodHook() {
                        @Override
                        protected void beforeHookedMethod(MethodHookParam param) {
                            log("Block call to getTemp()");
                        }

                        @Override
                        protected void afterHookedMethod(MethodHookParam param) {
                            int originalResult = (int) param.getResult();
                            int newTemperature = readTemperatureFromFile();

                            if (newTemperature > 0) {
                                if (DEBUG) log("getTemp - Original temperature: " + originalResult + ", new temperature: " + newTemperature);
                                param.setResult(newTemperature);
                            } else {
                                log("Failed to read temperature file, use original value: " + originalResult);
                            }
                        }
                    }
            );

            // Hook getThermalTemp 方法
            XposedHelpers.findAndHookMethod(
                    hwDataInterfaceClass,
                    "getThermalTemp",
                    int.class,
                    new XC_MethodHook() {
                        @Override
                        protected void beforeHookedMethod(MethodHookParam param) {
                            int type = (int) param.args[0];
                            if (DEBUG) log("Blocked getThermalTemp(), type: " + type);
                        }

                        @Override
                        protected void afterHookedMethod(MethodHookParam param) {
                            int originalResult = (int) param.getResult();
                            int newTemperature = readTemperatureFromFile();

                            if (newTemperature > 0) {
                                if (DEBUG) log("getThermalTemp - Original: " + originalResult + ", new: " + newTemperature);
                                param.setResult(newTemperature);
                            }
                        }
                    }
            );

            log("Hook executed successfully.");

        } catch (Throwable t) {
            logError("Failed to hook ZUI game service!", t);
        }
    }

    private void hookLenovoGamingService(XC_LoadPackage.LoadPackageParam lpparam) {
        try {
            // 尝试Hook联想游戏服务中的温度相关方法
            Class<?> thermalManagerClass = XposedHelpers.findClassIfExists(
                    "com.lenovo.gamingservice.ThermalManager",
                    lpparam.classLoader
            );

            if (thermalManagerClass != null) {
                XposedHelpers.findAndHookMethod(
                        thermalManagerClass,
                        "getCurrentTemperature",
                        new XC_MethodHook() {
                            @Override
                            protected void afterHookedMethod(MethodHookParam param) {
                                int originalResult = (int) param.getResult();
                                int newTemperature = readTemperatureFromFile();

                                if (newTemperature > 0) {
                                    if (DEBUG) log("Lenovo Game Service - temperature fix: "
                                            + originalResult
                                            + " -> "
                                            + newTemperature);
                                    param.setResult(newTemperature);
                                }
                            }
                        }
                );
                log("Hook for gaming service executed successfully.");
            } else {
                log("Unable to find class ThermalManager");
            }

        } catch (Throwable t) {
            logError("Hook failed!", t);
        }
    }

    private void hookGenericGamingService(XC_LoadPackage.LoadPackageParam lpparam) {
        try {
            // 尝试Hook通用的温度读取方法
            String[] temperatureMethods = {
                    "getTemperature",
                    "getCPUTemperature",
                    "getGPUTemperature",
                    "getThermalValue"
            };

            for (String methodName : temperatureMethods) {
                try {
                    XposedHelpers.findAndHookMethod(
                            "android.os.SystemProperties",
                            lpparam.classLoader,
                            methodName,
                            new XC_MethodHook() {
                                @Override
                                protected void afterHookedMethod(MethodHookParam param) {
                                    int newTemperature = readTemperatureFromFile();
                                    if (newTemperature > 0) {
                                        if (DEBUG) log("Generic temperature detection method "
                                                + methodName
                                                + " fixed, new temperature: "
                                                + newTemperature);
                                        param.setResult(newTemperature);
                                    }
                                }
                            }
                    );
                } catch (Throwable t) {
                    // 方法不存在是正常的，继续尝试下一个
                }
            }

            if (DEBUG) log("Generic detection method hook executed successfully.");

        } catch (Throwable t) {
            logError("Failed to hook generic temperature detection method", t);
        }
    }

    /**
     * 从 thermal_zone9 文件读取温度
     * @return 温度值（毫摄氏度），读取失败返回 -1
     */
    private int readTemperatureFromFile() {
        File thermalFile = new File(THERMAL_FILE_PATH);

        if (!thermalFile.exists()) {
            if (DEBUG) log("Temperature file does not exist: " + THERMAL_FILE_PATH);
            // 尝试其他可能的thermal文件路径
            return tryAlternativeThermalFiles();
        }

        if (!thermalFile.canRead()) {
            log("Failed to read file: permission denied " + THERMAL_FILE_PATH);
            return -1;
        }

        try (BufferedReader reader = new BufferedReader(new FileReader(thermalFile))) {
            String line = reader.readLine();

            if (line != null && !line.trim().isEmpty()) {
                int temperature = Integer.parseInt(line.trim());
                if (DEBUG) log("Read temperature data from file: " + temperature);
                return temperature;
            }

        } catch (IOException e) {
            logError("IO exception happened when reading temperature file", e);
        } catch (NumberFormatException e) {
            logError("Invalid temperature file format", e);
        }

        return -1;
    }

    /**
     * 尝试其他可能的thermal文件路径
     */
    private int tryAlternativeThermalFiles() {
        String[] alternativePaths = {
                "/sys/class/thermal/thermal_zone0/temp",
                "/sys/class/thermal/thermal_zone1/temp",
                "/sys/class/thermal/thermal_zone2/temp",
                "/sys/devices/virtual/thermal/thermal_zone9/temp",
                "/sys/class/hwmon/hwmon0/temp1_input"
        };

        for (String path : alternativePaths) {
            File thermalFile = new File(path);
            if (thermalFile.exists() && thermalFile.canRead()) {
                if (DEBUG) log("Alternate temperature file found: " + path);
                return readFromSpecificFile(path);
            }
        }

        log("Unable to find a valid temperature file.");
        return -1;
    }

    private int readFromSpecificFile(String filePath) {
        File thermalFile = new File(filePath);
        try (BufferedReader reader = new BufferedReader(new FileReader(thermalFile))) {
            String line = reader.readLine();

            if (line != null && !line.trim().isEmpty()) {
                return Integer.parseInt(line.trim());
            }
        } catch (Exception e) {
            logError("Failed to read temperature file: " + filePath, e);
        }
        // 忽略关闭异常
        return -1;
    }
}

File: app/src/main/java/com/qimian233/ztool/hook/modules/gametool/SocTemperatureFix.kt
package com.qimian233.ztool.hook.modules.gametool

import com.qimian233.ztool.data.keys.PreferenceKeys
import com.qimian233.ztool.data.keys.ScopeKeys
import com.qimian233.ztool.hook.base.AppHookModule
import io.github.libxposed.api.XposedModuleInterface.PackageLoadedParam
import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.IOException

/**
 * SOC temperature fix hook module.
 * Function: intercepts the temperature reading methods of the game service
 * and obtains the real temperature value from the thermal_zone9 file.
 */
class SocTemperatureFix : AppHookModule() {
    override fun getModuleName(): String = PreferenceKeys.FIX_SOC_TEMP.name

    override fun getTargetPackages(): Array<String> = arrayOf(ScopeKeys.GAME_SERVICE.packageName)

    override fun handleLoadPackage(param: PackageLoadedParam) {
        val classLoader = param.defaultClassLoader
        val packageName = param.packageName
        logger.debug("SocTemperatureFix: start processing package $packageName")
        hookZuiGameService(classLoader)
    }

    private fun hookZuiGameService(classLoader: ClassLoader) {
        try {
            val hwDataInterfaceClass =
                classLoader.loadClass("com.zui.game.service.util.HWDataInterface")

            // Hook the getTemp method
            val getTempMethod = hwDataInterfaceClass.getDeclaredMethod("getTemp")
            hookWithId(getTempMethod, "get_temp") { chain ->
                logger.info("Block call to getTemp()")
                val originalResult = chain.proceed() as Int
                val newTemperature = readTemperatureFromFile()
                if (newTemperature > 0) {
                    logger.trace("getTemp - Original temperature: $originalResult, new temperature: $newTemperature")
                    return@hookWithId newTemperature
                } else {
                    logger.warn("Failed to read temperature file, use original value: $originalResult")
                    return@hookWithId originalResult
                }
            }

            // Hook the getThermalTemp method
            val getThermalTempMethod = hwDataInterfaceClass.getDeclaredMethod(
                "getThermalTemp",
                Int::class.javaPrimitiveType
            )
            hookWithId(
                getThermalTempMethod,
                "get_thermal_temp"
            ) { chain ->
                val type = chain.getArg(0) as Int
                logger.debug("Blocked getThermalTemp(), type: $type")
                val originalResult = chain.proceed() as Int
                val newTemperature = readTemperatureFromFile()

                if (newTemperature > 0) {
                    logger.trace("getThermalTemp - Original: $originalResult, new: $newTemperature")
                    return@hookWithId newTemperature
                }
                originalResult
            }

            logger.info("Hook executed successfully.")
        } catch (t: Throwable) {
            logger.error("Failed to hook ZUI game service!", t)
        }
    }

    /**
     * Read the temperature from the thermal_zone9 file.
     * @return temperature in milli-Celsius, or -1 when reading fails
     */
    private fun readTemperatureFromFile(): Int {
        val thermalFile = File(THERMAL_FILE_PATH)

        if (!thermalFile.exists()) {
            logger.warn("Temperature file does not exist: $THERMAL_FILE_PATH")
            // Try other possible thermal file paths
            return tryAlternativeThermalFiles()
        }

        if (!thermalFile.canRead()) {
            logger.warn("Failed to read file: permission denied $THERMAL_FILE_PATH")
            return -1
        }

        try {
            BufferedReader(FileReader(thermalFile)).use { reader ->
                val line = reader.readLine()
                if (line != null && !line.trim { it <= ' ' }.isEmpty()) {
                    val temperature = line.trim { it <= ' ' }.toInt()
                    logger.debug("Read temperature data from file: $temperature")
                    return temperature
                }
            }
        } catch (e: IOException) {
            logger.error("IO exception happened when reading temperature file", e)
        } catch (e: NumberFormatException) {
            logger.error("Invalid temperature file format", e)
        }

        return -1
    }

    /**
     * Try other possible thermal file paths.
     */
    private fun tryAlternativeThermalFiles(): Int {
        val alternativePaths = arrayOf(
            "/sys/class/thermal/thermal_zone0/temp",
            "/sys/class/thermal/thermal_zone1/temp",
            "/sys/class/thermal/thermal_zone2/temp",
            "/sys/devices/virtual/thermal/thermal_zone9/temp",
            "/sys/class/hwmon/hwmon0/temp1_input"
        )

        for (path in alternativePaths) {
            val thermalFile = File(path)
            if (thermalFile.exists() && thermalFile.canRead()) {
                logger.debug("Alternate temperature file found: $path")
                return readFromSpecificFile(path)
            }
        }

        logger.warn("Unable to find a valid temperature file.")
        return -1
    }

    private fun readFromSpecificFile(filePath: String): Int {
        val thermalFile = File(filePath)
        try {
            BufferedReader(FileReader(thermalFile)).use { reader ->
                val line = reader.readLine()
                if (line != null && !line.trim { it <= ' ' }.isEmpty()) {
                    return line.trim { it <= ' ' }.toInt()
                }
            }
        } catch (e: Exception) {
            logger.error("Failed to read temperature file: $filePath", e)
        }
        // Ignore close exceptions
        return -1
    }

    companion object {
        private const val THERMAL_FILE_PATH = "/sys/class/thermal/thermal_zone9/temp"
    }
}
